File: server/model/event.go
package model

import (
	"sync"
	"time"
)

type Event struct {
	Timestamp     uint64                 `json:"timestamp"`
	Category      string                 `json:"category"`
	EventType     string                 `json:"event_type"`
	PID           uint32                 `json:"pid"`
	PPID          uint32                 `json:"ppid"`
	UID           uint32                 `json:"uid"`
	GID           uint32                 `json:"gid"`
	Comm          string                 `json:"comm"`
	ContainerID   string                 `json:"container_id"`
	ContainerName string                 `json:"container_name"`
	Filename      string                 `json:"filename,omitempty"`
	OldFilename   string                 `json:"old_filename,omitempty"`
	ExitCode      uint32                 `json:"exit_code,omitempty"`
	Mode          uint32                 `json:"mode,omitempty"`
	Flags         uint32                 `json:"flags,omitempty"`
	OldUID        uint32                 `json:"old_uid,omitempty"`
	SrcAddr       string                 `json:"src_addr,omitempty"`
	DstAddr       string                 `json:"dst_addr,omitempty"`
	SrcPort       uint16                 `json:"src_port,omitempty"`
	DstPort       uint16                 `json:"dst_port,omitempty"`
	Protocol      uint8                  `json:"protocol,omitempty"`
	WriteBytes    uint32                 `json:"write_bytes,omitempty"`
	Extra         map[string]interface{} `json:"extra,omitempty"`
}

type Stats struct {
	TotalEvents      uint64            `json:"total_events"`
	EventsPerSec     float64           `json:"events_per_sec"`
	ByCategory       map[string]uint64 `json:"by_category"`
	ByType           map[string]uint64 `json:"by_type"`
	ConnectedClients int               `json:"connected_clients"`
	LastEventTime    time.Time         `json:"last_event_time"`
}

type EventStore struct {
	mu         sync.RWMutex
	events     []Event
	capacity   int
	totalCount uint64
	byCategory map[string]uint64
	byType     map[string]uint64
	lastTime   time.Time
}

func NewEventStore(capacity int) *EventStore {
	return &EventStore{
		events:     make([]Event, 0, capacity),
		capacity:   capacity,
		byCategory: make(map[string]uint64),
		byType:     make(map[string]uint64),
	}
}

func (s *EventStore) Add(e Event) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.events) >= s.capacity {
		s.events = s.events[1:]
	}
	s.events = append(s.events, e)
	s.totalCount++
	s.byCategory[e.Category]++
	s.byType[e.EventType]++
	s.lastTime = time.Now()
}

func (s *EventStore) AddBatch(events []Event) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, e := range events {
		if len(s.events) >= s.capacity {
			s.events = s.events[1:]
		}
		s.events = append(s.events, e)
		s.totalCount++
		s.byCategory[e.Category]++
		s.byType[e.EventType]++
	}
	s.lastTime = time.Now()
}

func (s *EventStore) Query(category, eventType string, pid uint32, since uint64, limit int) []Event {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if limit <= 0 { limit = 1000 }
	result := make([]Event, 0, limit)
	for i := len(s.events) - 1; i >= 0; i-- {
		e := s.events[i]
		if category != "" && e.Category != category { continue }
		if eventType != "" && e.EventType != eventType { continue }
		if pid != 0 && e.PID != pid { continue }
		if since != 0 && e.Timestamp < since { continue }
		result = append(result, e)
		if len(result) >= limit { break }
	}
	return result
}

func (s *EventStore) Stats(clientCount int) Stats {
	s.mu.RLock()
	defer s.mu.RUnlock()
	byCat := make(map[string]uint64)
	byTyp := make(map[string]uint64)
	for k, v := range s.byCategory { byCat[k] = v }
	for k, v := range s.byType { byTyp[k] = v }
	var eps float64
	if !s.lastTime.IsZero() {
		elapsed := time.Since(s.lastTime).Seconds()
		if elapsed > 0 && elapsed < 60 { eps = float64(s.totalCount) / elapsed }
	}
	return Stats{
		TotalEvents: s.totalCount, EventsPerSec: eps,
		ByCategory: byCat, ByType: byTyp,
		ConnectedClients: clientCount, LastEventTime: s.lastTime,
	}
}

File: frontend/src/types/event.ts
export interface SecurityEvent {
  timestamp: number
  category: 'process' | 'network' | 'file'
  event_type: string
  pid: number
  ppid: number
  uid: number
  gid: number
  comm: string
  container_id: string
  container_name: string
  filename?: string
  old_filename?: string
  exit_code?: number
  mode?: number
  flags?: number
  old_uid?: number
  src_addr?: string
  dst_addr?: string
  src_port?: number
  dst_port?: number
  protocol?: number
  write_bytes?: number
}

export interface EventStats {
  total_events: number
  events_per_sec: number
  by_category: Record<string, number>
  by_type: Record<string, number>
  connected_clients: number
  last_event_time: string
}

export type EventCategory = 'process' | 'network' | 'file'

export const EVENT_TYPE_LABELS: Record<string, string> = {
  process_create: '进程创建', process_exit: '进程退出', privilege_escalation: '权限提升',
  tcp_connect: 'TCP连接', tcp_accept: 'TCP接受', tcp_close: 'TCP关闭',
  udp_send: 'UDP发送', udp_recv: 'UDP接收',
  file_create: '文件创建', file_modify: '文件修改', file_delete: '文件删除',
  file_rename: '文件重命名', file_chmod: '权限修改', file_chown: '所有者修改',
}

export const CATEGORY_COLORS: Record<string, string> = {
  process: '#3b82f6', network: '#10b981', file: '#f59e0b',
}

export const EVENT_TYPE_COLORS: Record<string, string> = {
  privilege_escalation: '#ef4444', process_create: '#3b82f6', process_exit: '#6366f1',
  tcp_connect: '#10b981', tcp_accept: '#059669', tcp_close: '#6b7280',
  udp_send: '#14b8a6', udp_recv: '#0d9488',
  file_create: '#f59e0b', file_modify: '#d97706', file_delete: '#dc2626',
  file_rename: '#8b5cf6', file_chmod: '#ec4899', file_chown: '#f97316',
}
